README.md
taskfarm
========
This package solves the problem of managing a loosely coupled taskfarm where
there are many tasks and the workers are entitrly independent of each other.
Instead of using a farmer process a database is used to hand out new tasks to
the workers. The workers contact a web application via http(s) to get a new
task.

You can use the [taskfarm-worker](https://github.com/mhagdorn/taskfarm-worker)
to connect to the taskfarm service.

Setup
-----
After installing the python package you need to connect to a database. For
testing purposes you can use sqlite. However, sqlite does not allow row
locking so if you use parallel workers a task may get assigned to the multiple
workers.

You can set the environment variable DATABASE_URL to configure the data base
connection. For example
```
export DATABASE_URL=sqlite:///app.db
```
or
```
export DATABASE_URL=postgresql://user:pw@host/db
```

You then need to create the tables by running
```
adminTF --init-db
```
You can then create some users
```
adminTF -u some_user -p some_password
```
These users are used by the worker to connect to the service.

![test package taskfarm](https://github.com/mhagdorn/taskfarm/workflows/test%20package%20taskfarm/badge.svg)


ChangeLog
2020-03-25 Magnus Hagdorn
 * taskfarm/routes.py: update how json data is extracted from response

2020-03-25 Magnus Hagdorn
 * taskfarm/routes.py: catch and handle exception

2020-03-25 Magnus Hagdorn
 * tests/test_taskfarm.py: add more tests

2020-03-25 Magnus Hagdorn
 * taskfarm/routes.py: remove unnecessary test

2020-03-25 Magnus Hagdorn
 * setup.py: use nose2 for testing, require newer version of flask
 * .gitignore: ignore some more stuff
 * tests/test_taskfarm.py: add some tests

2018-12-10 Magnus Hagdorn
 * taskfarm/models.py: add column to runs table handling the next task
 * taskfarm/routes.py: do not prepopulate tasks table, generate tasks on the 
   fly

2018-12-05 Magnus Hagdorn
 * taskfarm/__init__.py: setup logging when not running stand-alone
 * taskfarm/manage.py: fix bug
 * taskfarm/routes.py: make it work with older version of flask

2018-11-28 Magnus Hagdorn
 * setup.py: rename admin script
 * taskfarm/models.py: add some properties to describe the state of a run
 * taskfarm/routes.py: add routes to get/set state of runs
 * README.md: updated

2018-11-28 Magnus Hagdorn 
 * create ChangeLog, first working version


tests/test_taskfarm.py
from flask_testing import TestCase

from taskfarm import app, db
from taskfarm.models import User
import base64
from uuid import uuid1

user = 'test'
passwd = 'test'

headers = {}
headers['Authorization'] = 'Basic ' + base64.b64encode(
    (user + ':' + passwd).encode('utf-8')).decode('utf-8')


class TaskfarmTest(TestCase):

    def create_app(self):
        app.config['TESTING'] = True
        app.config['WTF_CSRF_ENABLED'] = False
        app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite://"

        # pass in test configuration
        return app

    def setUp(self):
        self.app = app.test_client()

        db.create_all()
        # create user
        u = User(username=user)
        u.hash_password(passwd)
        db.session.add(u)
        db.session.commit()

    def tearDown(self):
        db.session.remove()
        db.drop_all()

    def test_user(self):
        u = 'test2'
        p = 'testing'
        u = User(username=u)
        u.hash_password(p)
        self.assertTrue(u.verify_password(p))
        self.assertFalse(u.verify_password(p+'not'))

        db.session.add(u)
        db.session.commit()

        assert u in db.session

    def test_get_auth_token(self):
        response = self.app.get('/api/token',  headers=headers)
        self.assertEqual(response.status_code,  200)

    def test_auth_with_token(self):
        response = self.app.get('/api/token',  headers=headers)
        token = response.get_json()['token']

        h = {}
        h['Authorization'] = 'Basic ' + base64.b64encode(
            token.encode('utf-8')+b': ').decode('utf-8')
        h['Content-Type'] = 'application/json'
        h['Accept'] = 'application/json'

        response = self.app.get('/api/token',  headers=h)
        self.assertEqual(response.status_code,  200)

    def create_run(self, numTasks):
        response = self.app.post('/api/run',
                                 headers=headers,
                                 json={'numTasks': numTasks})
        self.assertEqual(response.status_code,  201)
        return response.get_json()

    def create_worker(self, hostname, pid):
        data = {'uuid':  uuid1().hex,
                'hostname':  hostname,
                'pid': pid}
        response = self.app.post('/api/worker',
                                 headers=headers,
                                 json=data)
        self.assertEqual(response.status_code,  201)
        return response.get_json()

    def get_run_info(self, run_uuid, info=None, status=200):
        data = {}
        if info is not None:
            data['info'] = info
        response = self.app.get('/api/runs/'+run_uuid,
                                headers=headers,  query_string=data)
        self.assertEqual(response.status_code,  status)
        return response.get_json()

    def restart_run(self, run_uuid, all=None, status=204):
        data = {}
        if all is not None:
            data['all'] = all
        response = self.app.post('/api/runs/{}/restart'.format(run_uuid),
                                 headers=headers,  query_string=data)
        self.assertEqual(response.status_code,  status)

    def get_task(self, run_uuid, worker_uuid, status=201):
        response = self.app.post('/api/runs/{}/task'.format(run_uuid),
                                 headers=headers,
                                 json={'worker_uuid': worker_uuid})
        self.assertEqual(response.status_code,  status)
        return response.get_json()

    def get_task_info(self, run, task, info='', status=200):
        args = {}
        if len(info) > 0:
            args['info'] = info
        response = self.app.get('/api/runs/{}/tasks/{}'.format(run, task),
                                headers=headers,
                                query_string=args)
        self.assertEqual(response.status_code,  status)
        return response.get_json()

    def update_task_info(self, run, task, info, status=204):
        response = self.app.put('/api/runs/{}/tasks/{}'.format(run, task),
                                headers=headers,
                                json=info)
        self.assertEqual(response.status_code,  status)

    def test_create_run(self):
        nt = 10
        run = self.create_run(nt)
        self.assertEqual(run['numTasks'], nt)

    def test_get_all_runs(self):
        numTasks = [10, 20, 30]
        runs = []
        for nt in numTasks:
            run = self.create_run(nt)
            runs.append(run)
        response = self.app.get('/api/runs', headers=headers)
        self.assertEqual(response.status_code,  200)
        data = response.get_json()['data']
        for i in range(len(runs)):
            self.assertDictEqual(runs[i], data[i])

    def test_get_run(self):
        nt = 10
        run = self.create_run(nt)

        response = self.app.get('/api/runs/'+run['uuid'],  headers=headers)
        self.assertEqual(response.status_code,  200)
        data = response.get_json()
        self.assertEqual(data['uuid'], run['uuid'])
        self.assertEqual(data['numTasks'], nt)
        for k in ['numComputing', 'numDone', 'numWaiting', 'percentDone']:
            self.assertEqual(data[k], 0)

    def test_get_run_na(self):
        response = self.app.get('/api/runs/no_such_run',  headers=headers)
        self.assertEqual(response.status_code,  404)

    def test_get_run_info_all(self):
        nt = 10
        run = self.create_run(nt)
        rinfo = self.get_run_info(run['uuid'])
        for k in run:
            self.assertEqual(run[k], rinfo[k])

    def test_get_run_info(self):
        nt = 10
        run = self.create_run(nt)
        for i in ['percentDone', 'numWaiting', 'numDone', 'numComputing']:
            self.get_run_info(run['uuid'], info=i)
        self.get_run_info(run['uuid'], info='blub', status=404)

    def test_delete_run(self):
        run = self.create_run(10)
        response = self.app.delete('/api/runs/'+run['uuid'],  headers=headers)
        self.assertEqual(response.status_code,  204)

    def test_delete_run_na(self):
        response = self.app.delete('/api/runs/no_such_run',  headers=headers)
        self.assertEqual(response.status_code,  404)

    def test_create_worker(self):
        self.create_worker('blub', 1000)

    def test_get_task_noworker(self):
        nt = 10
        run = self.create_run(nt)
        response = self.app.post('/api/{}/task'.format(run['uuid']),
                                 headers=headers)
        self.assertEqual(response.status_code,  404)

    def test_get_task_wrongworker(self):
        nt = 10
        run = self.create_run(nt)
        self.get_task(run['uuid'], 'no_worker', status=404)

    def test_get_task_norun(self):
        worker = self.create_worker('blub', 1000)
        self.get_task('norun', worker['uuid'], status=404)

    def test_get_task(self):
        nt = 10
        run = self.create_run(nt)
        worker = self.create_worker('blub', 1000)
        # create all tasks
        for i in range(nt):
            task = self.get_task(run['uuid'], worker['uuid'])
            self.assertEqual(task['percentCompleted'], 0)
            self.assertEqual(task['status'], 'computing')
            self.assertEqual(task['task'], i)
        # and another one which should indicate all tasks are handed out
        task = self.get_task(run['uuid'], worker['uuid'], status=204)

        # all tasks should have status computing and
        # there should be no waiting tasks
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numComputing'], nt)
        self.assertEqual(rinfo['numWaiting'], 0)

    def test_get_task_info_norun(self):
        self.get_task_info('norun', 1, status=404)

    def test_get_task_info_outofrange(self):
        nt = 10
        run = self.create_run(nt)
        self.get_task_info(run['uuid'], -1, status=404)
        self.get_task_info(run['uuid'], nt, status=404)

    def test_get_task_info_wronginfo(self):
        nt = 10
        run = self.create_run(nt)
        self.get_task_info(run['uuid'], 0,
                           info='no_such_thing', status=404)

    def test_get_existing_task_info(self):
        nt = 10
        run = self.create_run(nt)
        worker = self.create_worker('blub', 1000)
        task = self.get_task(run['uuid'], worker['uuid'])

        tinfo = self.get_task_info(run['uuid'], task['task'])
        self.assertDictEqual(task, tinfo)

    def test_get_task_info(self):
        nt = 10
        run = self.create_run(nt)
        # create new tasks on the fly
        tid = 0
        tinfo = self.get_task_info(run['uuid'], tid, status=200)
        self.assertEqual(tinfo['task'], tid)

        for i in ['status', 'percentCompleted']:
            ti = self.get_task_info(run['uuid'], tid, info=i, status=200)
            self.assertEqual(tinfo[i], ti[i])

    def test_update_task_info_nothing(self):
        nt = 10
        run = self.create_run(nt)
        self.update_task_info(run['uuid'], 0, None, status=400)

    def test_update_task_info(self):
        nt = 10
        run = self.create_run(nt)
        tid = 0

        for pc in [10, 20, 50.5, 100]:
            self.update_task_info(run['uuid'], tid, {'percentCompleted': pc})

        for s in ['waiting', 'computing', 'done']:
            self.update_task_info(run['uuid'], tid, {'status': s})

        self.update_task_info(run['uuid'], tid,
                              {'status': 'wrong'}, status=400)

    def test_get_task_waiting(self):
        nt = 10
        run = self.create_run(nt)
        worker = self.create_worker('blub', 1000)
        # create all tasks
        for i in range(nt):
            task = self.get_task(run['uuid'], worker['uuid'])

        # reset a particular task
        tid = 5
        self.update_task_info(run['uuid'], tid, {'status': 'waiting'})
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numComputing'], nt-1)

        task = self.get_task(run['uuid'], worker['uuid'])
        self.assertEqual(task['task'], tid)

    def test_restart_run_norun(self):
        self.restart_run('no_run', status=404)

    def test_restart_run_wrongparam(self):
        nt = 10
        run = self.create_run(nt)
        self.restart_run(run['uuid'], all='wrong', status=404)

    def test_restart_run(self):
        nt = 10
        run = self.create_run(nt)
        for i in range(0, nt, 2):
            self.update_task_info(run['uuid'], i,
                                  {'status': 'done',
                                   'percentCompleted': 100})
        for i in range(1, nt, 2):
            self.update_task_info(run['uuid'], i,
                                  {'status': 'computing',
                                   'percentCompleted': 50})
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numWaiting'], 0)
        self.assertEqual(rinfo['numComputing'], nt/2)
        self.assertEqual(rinfo['numDone'], nt/2)

        self.restart_run(run['uuid'])
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numWaiting'], nt/2)
        self.assertEqual(rinfo['numComputing'], 0)
        self.assertEqual(rinfo['numDone'], nt/2)

    def test_restart_run_all(self):
        nt = 10
        run = self.create_run(nt)
        for i in range(0, nt, 2):
            self.update_task_info(run['uuid'], i,
                                  {'status': 'done',
                                   'percentCompleted': 100})
        for i in range(1, nt, 2):
            self.update_task_info(run['uuid'], i,
                                  {'status': 'computing',
                                   'percentCompleted': 50})
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numWaiting'], 0)
        self.assertEqual(rinfo['numComputing'], nt/2)
        self.assertEqual(rinfo['numDone'], nt/2)

        self.restart_run(run['uuid'], all='True')
        rinfo = self.get_run_info(run['uuid'])
        self.assertEqual(rinfo['numWaiting'], nt)
        self.assertEqual(rinfo['numComputing'], 0)
        self.assertEqual(rinfo['numDone'], 0)


taskfarm/models.py
__all__ = ['TaskState', 'User', 'Run', 'Task', 'Worker']

from .application import db, app
from itsdangerous import JSONWebSignatureSerializer as Serializer, BadSignature
from sqlalchemy import func
import enum
from passlib.apps import custom_app_context as pwd_context


class TaskState(enum.Enum):
    waiting = 1
    computing = 2
    done = 3


class User(db.Model):
    __tablename__ = 'users'
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(32), index=True)
    password_hash = db.Column(db.String(128))

    def __repr__(self):
        return '<User {}>'.format(self.username)

    def hash_password(self, password):
        self.password_hash = pwd_context.hash(password)

    def verify_password(self, password):
        return pwd_context.verify(password, self.password_hash)

    def generate_auth_token(self):
        pass
        s = Serializer(app.config['SECRET_KEY'])
        return s.dumps({'id': self.id})

    @staticmethod
    def verify_auth_token(token):
        s = Serializer(app.config['SECRET_KEY'])
        try:
            data = s.loads(token)
        except BadSignature:
            return None  # invalid token
        user = User.query.get(data['id'])
        return user


class Run(db.Model):
    __tablename__ = 'runs'

    id = db.Column(db.Integer, primary_key=True)
    uuid = db.Column(db.String, index=True, unique=True)
    nextTask = db.Column(db.Integer, default=0)
    numTasks = db.Column(db.Integer)

    tasks = db.relationship("Task", backref='run', lazy='dynamic')

    @property
    def to_dict(self):
        return {
            "id": self.id,
            "uuid": self.uuid,
            "numTasks": self.numTasks,
        }

    @property
    def full_status(self):
        info = self.to_dict
        for k in ['percentDone', 'numWaiting', 'numDone', 'numComputing']:
            info[k] = getattr(self, k)
        return info

    @property
    def numListedTasks(self):
        return Task.query.filter_by(run_id=self.id).count()

    @property
    def percentDone(self):
        try:
            return db.session.query(func.sum(Task.percentCompleted)) \
                             .filter_by(run_id=self.id).scalar()/self.numTasks
        except Exception:
            return 0.

    def runStatus(self, status):
        return Task.query.filter_by(run_id=self.id, status=status).count()

    @property
    def numWaiting(self):
        return self.runStatus(TaskState.waiting)

    @property
    def numDone(self):
        return self.runStatus(TaskState.done)

    @property
    def numComputing(self):
        return self.runStatus(TaskState.computing)


class Task(db.Model):
    __tablename__ = 'tasks'

    id = db.Column(db.Integer, primary_key=True)
    task = db.Column(db.Integer)
    status = db.Column(db.Enum(TaskState), default=TaskState.waiting)
    started = db.Column(db.DateTime)
    updated = db.Column(db.DateTime)
    percentCompleted = db.Column(db.Float, default=0.)

    run_id = db.Column(db.Integer, db.ForeignKey('runs.id'))
    worker_id = db.Column(db.Integer, db.ForeignKey('workers.id'))

    @property
    def to_dict(self):
        return {
            "id": self.id,
            "task": self.task,
            "percentCompleted": self.percentCompleted,
            "status": self.status.name,
            }


class Worker(db.Model):
    __tablename__ = 'workers'

    id = db.Column(db.Integer, primary_key=True)
    uuid = db.Column(db.String, index=True, unique=True)
    hostname = db.Column(db.String)
    pid = db.Column(db.Integer)
    start = db.Column(db.DateTime)

    tasks = db.relationship("Task", backref='worker', lazy='dynamic')

    def __repr__(self):
        return ('{uuid: {uuid}, hostname: {hostname}, ' +
                'pid: {pid}, start {start}}').format(self)


taskfarm/routes.py
from .application import app, db
from flask_httpauth import HTTPBasicAuth
from flask import jsonify, request, abort, g
from uuid import uuid4
from datetime import datetime
from .models import Worker, Task, Run, TaskState, User
import logging

auth = HTTPBasicAuth()


@auth.verify_password
def verify_password(username_or_token, password):
    # first try to authenticate by token
    user = User.verify_auth_token(username_or_token)
    if not user:
        # try to authenticate with username/password
        user = User.query.filter_by(username=username_or_token).first()
        if not user or not user.verify_password(password):
            return False
    g.user = user
    return True


@app.route('/api/token')
@auth.login_required
def get_auth_token():
    token = g.user.generate_auth_token()
    return jsonify({'token': token.decode('ascii')})


@app.route('/api/run', methods=['POST'])
@auth.login_required
def create_run():
    # create a new run
    if not request.get_json() or 'numTasks' not in request.get_json():
        abort(400)

    run = Run(uuid=uuid4().hex, numTasks=request.get_json()['numTasks'])
    db.session.add(run)
    db.session.commit()
    return jsonify(run.to_dict), 201


@app.route('/api/runs', methods=['GET'])
@auth.login_required
def get_all_runs():
    results = []
    for run in Run.query.all():
        results.append(run.to_dict)
    return jsonify({'data': results}), 200


@app.route('/api/runs/<string:uuid>', methods=['GET', 'DELETE'])
@auth.login_required
def get_run(uuid):
    run = Run.query.filter_by(uuid=uuid).first()
    if not run:
        logging.error('no run with uuid={}'.format(uuid))
        abort(404)

    if request.method == 'GET':
        info = request.args.get('info', '')
        if info == '':
            result = run.full_status
        elif info in ['percentDone', 'numWaiting',
                      'numDone', 'numComputing']:
            result = {info: getattr(run, info)}
        else:
            abort(404)
        return jsonify(result), 200
    elif request.method == 'DELETE':
        db.session.query(Task).filter_by(run_id=run.id).delete()
        db.session.delete(run)
        db.session.commit()
        return '', 204
    abort(500)


@app.route('/api/runs/<string:uuid>/restart', methods=['POST'])
@auth.login_required
def restart_tasks(uuid):
    run = Run.query.filter_by(uuid=uuid).first()
    if not run:
        logging.error('no run with uuid={}'.format(uuid))
        abort(404)
    restart_all = request.args.get('all', 'False')
    if restart_all == 'True':
        db.session.query(Task).filter_by(run_id=run.id) \
                              .update({'status': TaskState.waiting,
                                       'percentCompleted': 0.})
    elif restart_all == 'False':
        db.session.query(Task).filter(Task.run_id == run.id,
                                      Task.percentCompleted < 100) \
                              .update({'status': TaskState.waiting})
    else:
        abort(404)
    db.session.commit()
    return '', 204


@app.route('/api/runs/<string:uuid>/task', methods=['POST'])
@auth.login_required
def get_task(uuid):
    if not request.get_json() or 'worker_uuid' not in request.get_json():
        abort(400)
    worker_uuid = request.get_json()['worker_uuid']

    worker = Worker.query.filter_by(uuid=worker_uuid).first()
    if not worker:
        logging.error('no worker with uuid={}'.format(worker_uuid))
        abort(404)

    run = db.session.query(Run).with_for_update().filter_by(uuid=uuid).first()
    if not run:
        logging.error('no run with uuid={}'.format(uuid))
        db.session.rollback()
        abort(404)

    task = db.session.query(Task).with_for_update() \
                                 .filter_by(status=TaskState.waiting,
                                            run_id=run.id).first()
    if not task and run.nextTask < run.numTasks:
        # no waiting tasks, create the next one
        if run.numListedTasks == run.nextTask:
            task = Task(task=run.nextTask, run=run)
            run.nextTask += 1
        else:
            for tid in range(run.nextTask, run.numTasks):
                t = Task.query.filter_by(run_id=run.id, task=tid).first()
                if not t:
                    task = Task(task=run.nextTask, run=run)
                    break
            run.nextTask = t+1
        db.session.add(run)

    if task:
        task.status = TaskState.computing
        task.started = datetime.now()
        task.worker = worker
        db.session.add(task)
    db.session.commit()

    if task is None:
        return '', 204
    else:
        return jsonify(task.to_dict), 201


@app.route('/api/runs/<string:uuid>/tasks/<int:taskID>',
           methods=['GET', 'PUT'])
@auth.login_required
def taskInfo(uuid, taskID):
    run = Run.query.filter_by(uuid=uuid).first()
    if not run:
        logging.error('no run with uuid={}'.format(uuid))
        abort(404)
    task = Task.query.filter_by(run_id=run.id, task=taskID).first()
    if not task:
        if taskID < 0 or taskID >= run.numTasks:
            logging.error('no taskID {} outside range(0,{})'
                          .format(taskID, run.numTasks))
            abort(404)

        # create a new task
        task = Task(task=taskID, run=run)
        db.session.add(task)
        if run.nextTask == taskID:
            run.nextTask += 1
            db.session.add(run)
        db.session.commit()

    if request.method == 'GET':
        info = request.args.get('info', '')
        if info == '':
            result = task.to_dict
        elif info in ['status', 'percentCompleted']:
            result = {info: task.to_dict[info]}
        else:
            abort(404)
        return jsonify(result), 200
    elif request.method == 'PUT':
        if not request.get_json():
            abort(400)
        data = request.get_json()
        for info in data:
            try:
                setattr(task, info, data[info])
            except Exception as e:
                return str(e), 400
        task.updated = datetime.now()
        db.session.add(task)
        try:
            db.session.commit()
        except Exception as e:
            return str(e), 400
        return '', 204
    abort(500)


@app.route('/api/worker', methods=['POST'])
@auth.login_required
def create_worker():
    if not request.get_json():
        abort(400)
    data = request.get_json()
    for k in ['uuid', 'hostname', 'pid']:
        if k not in data:
            abort(400)
    worker = Worker(uuid=data['uuid'],
                    hostname=data['hostname'],
                    pid=data['pid'],
                    start=datetime.now()
                    )
    db.session.add(worker)
    db.session.commit()

    return jsonify({'uuid': worker.uuid, 'id': worker.id}), 201
